### PR TITLE
Add mask on nan

### PR DESCRIPTION
**Pull request by Metriida**
Allows you to calculate stats and train when the dataset contains nan with mask_on_nan option of trainer.

When calculating stats, nan are ignored.
During training, a mask is added to the inputs, which corresponds to the union of the nan in inputs and forcing.
In addition, loss and metrics are not calculated where there are nan on the target.

I also added batch in plots to have more flexibility in plots, e.g. compare input output target.

### CODE REVIEW

File: py4cast/datasets/compute_dataset_stats.py [hunk 1]
```diff
@@ -20,12 +21,13 @@ def compute_mean_std_min_max(
...
-    best_min = torch.min(flat_input, dim=0).values
-    best_max = torch.max(flat_input, dim=0).values
+    best_min = torch.min(torch.nan_to_num(flat_input, torch.inf), dim=0).values
+    best_max = torch.max(torch.nan_to_num(flat_input, -torch.inf), dim=0).values
 
-    if torch.isnan(best_min).any() or torch.isnan(best_max).any():
-        raise ValueError(
-            "Your dataset contain NaN values, which prevent the calculation of statistics."
+    if torch.isnan(flat_input).any():
+        flat_input = torch.nan_to_num(flat_input)
```
Comment (kalharko) [inline py4cast/datasets/compute_dataset_stats.py:28]: Explicit replace value.
```suggestion
        flat_input = torch.nan_to_num(flat_input, nan=0)
```

File: py4cast/datasets/compute_dataset_stats.py [hunk 2]
```diff
@@ -20,12 +21,13 @@ def compute_mean_std_min_max(
...
+        warnings.warn(
+            "Your dataset contain NaN values, statistics will be calculated ignoring the NaN."
```
Comment (kalharko) [inline py4cast/datasets/compute_dataset_stats.py:30]:
```suggestion
            "py4cast.datasets.compute_dataset_stats.compute_mean_std_min_max:\n\tYour dataset contain NaN values, statistics will be calculated ignoring the NaN."
```

**Review by kalharko — CHANGES_REQUESTED** [review]

File: py4cast/lightning.py [hunk 3]
```diff
@@ -674,10 +677,41 @@ def _next_x(
...
+
+            # Combine masks for forcing
+            mask = torch.isnan(forcing.tensor)
+            for i in range(mask.shape[-1]):
+                combined_mask = (
+                    combined_mask | mask[:, :, :, i]
+                )  # Union of size masks (batch, lat, lon)
+
+            mask_list.append(
+                ~combined_mask.unsqueeze(-1)
```
Comment (kalharko) [inline py4cast/lightning.py:702]: Comment use of uncommon operator `~`.
```suggestion
                ~combined_mask.unsqueeze(-1)  # unsqueeze and invert combined_mask
```

File: py4cast/lightning.py [hunk 4]
```diff
@@ -674,10 +677,41 @@ def _next_x(
...
+            )  # shape [(batch, lat, lon, param)]
+
+            # replace nan by 0 in inputs
+            inputs = [torch.nan_to_num(input) for input in inputs]
```
Comment (kalharko) [inline py4cast/lightning.py:706]: Explicit default value.
```suggestion
            inputs = [torch.nan_to_num(input, nan=0) for input in inputs]
```

File: py4cast/lightning.py [hunk 5]
```diff
@@ -674,10 +677,41 @@ def _next_x(
...
+            # replace nan by 0 in forcing
+            forcing.tensor = torch.nan_to_num(forcing.tensor)
```
Comment (kalharko) [inline py4cast/lightning.py:708]: Explicit default value.
```suggestion
            forcing.tensor = torch.nan_to_num(forcing.tensor, nan=0)
```

File: py4cast/lightning.py [hunk 6]
```diff
@@ -701,6 +735,20 @@ def mask_tensor(self, x):
...
+        """
+        Returns a mask matching the nan values in target, same shape as the target.
+        Replaces the nan values by zeros in the target.
+        """
+        if self.mask_on_nan:
+            mask = ~torch.isnan(target.tensor)
+            target.tensor = torch.nan_to_num(target.tensor)
+            prediction.tensor = torch.nan_to_num(prediction.tensor)
+            return mask
+        return torch.ones_like(target.tensor)
```
Comment (kalharko) [inline py4cast/lightning.py:750]: Does this function truly do what it's docstring say it does ?
What's the point of the unused `prediction: NamedTensor` parameter ?
```suggestion
    def get_mask_on_nan(self, target: NamedTensor) -> torch.Tensor:
        """Returns a mask matching the nan values in target, same shape as the target.
        Replaces the nan values by zeros in the target.
        """
        if self.mask_on_nan:
            return ~torch.isnan(target.tensor)
        target.tensor = torch.nan_to_num(target.tensor, nan=0)
        return torch.ones_like(target.tensor)
```

File: py4cast/lightning.py [hunk 7]
```diff
@@ -781,14 +832,16 @@ def on_validation_start(self):
                 ),
             ]
 
-    def validation_step(self, batch: ItemBatch, batch_idx):
+    def validation_step(self, batch: ItemBatch, batch_idx: int):
         """
         Run validation on single batch
         """
```
Comment (kalharko) [inline py4cast/lightning.py:838]: Docstring formating
```suggestion
        """Runs validation on a single batch
        """
```

File: py4cast/lightning.py [hunk 8]
```diff
@@ -810,14 +863,34 @@ def validation_step(self, batch: ItemBatch, batch_idx):
...
+    def validation_step_logging(
+        self,
+        batch: ItemBatch,
+        prediction: NamedTensor,
+        target: NamedTensor,
+        mask: torch.Tensor,
+    ):
+        """
+        Save metrics and plots of validation to the tensorboard
+        """
```
Comment (kalharko) [inline py4cast/lightning.py:877]: Docstring formating.
```suggestion
        """Saves metrics and plots of validation to the tensorboard
        """
```

File: py4cast/lightning.py [hunk 9]
```diff
@@ -893,14 +966,16 @@ def on_test_start(self):
                 ),
             ]
 
-    def test_step(self, batch: ItemBatch, batch_idx):
+    def test_step(self, batch: ItemBatch, batch_idx: int):
         """
         Run test on single batch
         """
```
Comment (kalharko) [inline py4cast/lightning.py:972]: Docstring formating.
```suggestion
        """Runs test on single batch
        """
```

File: py4cast/lightning.py [hunk 10]
```diff
@@ -918,14 +993,30 @@ def test_step(self, batch: ItemBatch, batch_idx):
...
+    def test_step_logging(
+        self,
+        batch: ItemBatch,
+        prediction: NamedTensor,
+        target: NamedTensor,
+        mask: torch.Tensor,
+    ):
+        """
+        Save metrics and plots of test to the tensorboard
+        """
```
Comment (kalharko) [inline py4cast/lightning.py:1007]: Docstring formating.
```suggestion
        """Saves metrics and plots of test to the tensorboard
        """
```

**Review by kalharko — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 11]
```diff
@@ -6,7 +6,7 @@ networkx==3.3
 Cartopy==0.24.1
 tueplots==0.0.17
 cfgrib==0.9.14.1
-xarray==2024.9.0
+xarray>=2024.9.0
```
Comment (kalharko) [inline requirements.txt:9]: Please explain why this change was necessary in PR description.

File: py4cast/plots.py
Comment (LBerth) [inline py4cast/plots.py:1]: Would it be possible to add the input data as argument for the plots instead of the batch ? because in the batch, potentially you have data about multiple samples of the batch (depending on the batch size), but the prediction and target variables describe only one sample. So you will pass as argument a lot of data that you don't need.

**Review by kalharko — APPROVED** [review]
